source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -47,171 +51,235 @@ GlslFragmentGenerator::GlslFragmentGenerator() :
+        for (size_t i = 0; i < publicUniforms.size(); ++i)
         {
-            emitString(COMMA, stage);
+            const ShaderPort* shaderPort = publicUniforms[i];
+
+            if (shaderPort->getType() == Type::MATRIX33)
+                convertMatrixStrings.push_back(publicUniforms[i]->getVariable());
+
+            emitArgument(shaderPort);
         }
-        emitLineEnd(stage, false);
-    }
-    // Special case handling of world space normals which for now must be added 
-    // as a "dummy" argument if it exists.
-    const VariableBlock& streams = stage.getInputBlock(HW::VERTEX_DATA);
-    const ShaderPort* port = streams.find(HW::T_NORMAL_WORLD);
-    if (port)
-    { 
-        emitLineBegin(stage);
-        emitString(COMMA, stage);
-        emitVariableDeclaration(port, EMPTY_STRING, context, stage, false);
-        emitLineEnd(stage, false);
-    }
 
-    if (context.getOptions().hwTransparency)
-    {
-        // A dummy argument not used in the generated shader code but necessary to
-        // map onto an OGS fragment parameter and a shading node DG attribute with
-        // the same name that can be set to a non-0 value to let Maya know that the
-        // surface is transparent.
-        emitLineBegin(stage);
-        emitString(COMMA, stage);
-        emitString("float ", stage);
-        emitString(OgsXmlGenerator::VP_TRANSPARENCY_NAME, stage);
-        emitLineEnd(stage, false);
-    }
+        for (size_t i = 0; i < vertexData.size(); ++i)
+            emitArgument(vertexData[i]);
 
-    emitScopeEnd(stage);
+        if (context.getOptions().hwTransparency)
+        {
+            // A dummy argument not used in the generated shader code but necessary to
+            // map onto an OGS fragment parameter and a shading node DG attribute with
+            // the same name that can be set to a non-0 value to let Maya know that the
+            // surface is transparent.
+
+            if (!firstArgument)
+            {
+                emitString(COMMA, pixelStage);
+                emitLineEnd(pixelStage, false);
+            }
+
+            emitLineBegin(pixelStage);
+            emitString("float ", pixelStage);
+            emitString(OgsXmlGenerator::VP_TRANSPARENCY_NAME, pixelStage);
+            emitLineEnd(pixelStage, false);
+        }
+    }
+    emitScopeEnd(pixelStage);
 
     // Add function body
-    emitScopeBegin(stage);
+    emitScopeBegin(pixelStage);
 
     if (graph.hasClassification(ShaderNode::Classification::CLOSURE))
     {
         // Handle the case where the graph is a direct closure.
         // We don't support rendering closures without attaching 
         // to a surface shader, so just output black.
-        emitLine("return vec3(0.0)", stage);
+        emitLine("return vec3(0.0)", pixelStage);
     }
     else
     {
+        // Copy vertex data values from the root function's arguments to a
```

Comment:
Where vertex data is copied to a global variable to mimic GLSL input blocks in HLSL.